analyzers/src/SonarAnalyzer.CSharp/Rules/RedundancyInConstructorDestructorDeclaration.cs

```diff
@@ -89,32 +88,34 @@ private static void CheckConstructorDeclaration(SonarSyntaxNodeReportingContext
             constructorDeclaration is { ParameterList.Parameters.Count: 0, Body.Statements.Count: 0 }
-            && (IsSinglePublicConstructor(constructorDeclaration, semanticModel)
+            && !IsStructWithInitializedFieldOrProperty(constructorDeclaration.Parent, typeSymbol)
+            && (IsSinglePublicConstructor(constructorDeclaration, typeSymbol)
                 || constructorDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword));
 
-        private static bool IsSinglePublicConstructor(ConstructorDeclarationSyntax constructorDeclaration, SemanticModel semanticModel) =>
+        private static bool IsStructWithInitializedFieldOrProperty(SyntaxNode parentType, Lazy<INamedTypeSymbol> typeSymbol) =>
+            parentType.IsAnyKind(SyntaxKind.StructDeclaration, SyntaxKindEx.RecordStructDeclaration)
+            && ContainsInitializedFieldOrProperty(typeSymbol.Value);
+
+        private static bool IsSinglePublicConstructor(ConstructorDeclarationSyntax constructorDeclaration, Lazy<INamedTypeSymbol> typeSymbol) =>
             constructorDeclaration.Modifiers.Any(SyntaxKind.PublicKeyword)
             && IsInitializerEmptyOrRedundant(constructorDeclaration.Initializer)
-            && TypeHasExactlyOneConstructor(constructorDeclaration, semanticModel);
+            && TypeHasExactlyOneConstructor(typeSymbol.Value);
 
         private static bool IsInitializerEmptyOrRedundant(ConstructorInitializerSyntax initializer) =>
             initializer is null
             || (initializer.ArgumentList.Arguments.Count == 0
                 && initializer.ThisOrBaseKeyword.IsKind(SyntaxKind.BaseKeyword));
 
-        private static bool TypeHasExactlyOneConstructor(BaseMethodDeclarationSyntax constructorDeclaration, SemanticModel semanticModel)
-        {
-            var symbol = semanticModel.GetDeclaredSymbol(constructorDeclaration);
-            return symbol != null
-                   && symbol.ContainingType
-                            .GetMembers()
-                            .OfType<IMethodSymbol>()
-                            .Count(m => m.MethodKind == MethodKind.Constructor && !m.IsImplicitlyDeclared) == 1;
-        }
+        private static bool TypeHasExactlyOneConstructor(INamedTypeSymbol typeSymbol) =>
+            typeSymbol != null
+            && typeSymbol
+                .GetMembers()
+                .OfType<IMethodSymbol>()
+                .Count(m => m.MethodKind == MethodKind.Constructor && !m.IsImplicitlyDeclared) == 1;
 
-        private static bool ContainsInitializedFieldOrProperty(INamedTypeSymbol symbol) =>
-            symbol.GetMembers().Any(f => f.DeclaringSyntaxReferences.Any(d => d.GetSyntax().GetInitializer() is not null));
+        private static bool ContainsInitializedFieldOrProperty(INamedTypeSymbol typeSymbol) =>
+            typeSymbol != null && typeSymbol.GetMembers().Any(f => f.DeclaringSyntaxReferences.Any(d => d.GetSyntax().GetInitializer() is not null));
```

Comment:
```suggestion
        private static bool ContainsInitializedFieldOrProperty(INamedTypeSymbol typeSymbol) =>
            typeSymbol != null && typeSymbol.GetMembers().Any(f => f.Kind is SymbolKind.Field or SymbolKind.Property && f.DeclaringSyntaxReferences.Any(d => d.GetSyntax().GetInitializer() is not null));
```

Reply:
Another option would be to use `typeSymbol.DeclaringSyntaxReferences` and start from there. This should give you all `BaseTypeDeclarationSyntaxNode` of the partial type declaration and you can do a syntax-based search from there.